File: HouzLink/HouzLink/Model/Device.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace HouzLink.Model
{
    public class CommandResult
    {
        public ResultEnm Result { get; set; }
        public Device Device { get; set; }

        public override string ToString()
        {
            return $"{Result}: {Device}";
        }
    }

    public enum ResultEnm
    {
        SentOk = 1,
        SentRetry = 2,
        SentFail = 3,
        Received = 4
    }



    public class Device
    {
        public Device()
        {
        }

        //MSG to be sent: N1DA12ABCD
        //0123456789

        //N1DA12ABCD < query temp
        //N1DC330001 < offlice light on
        //N1DC330000 < offlice light off

        //[id:0x12|cmd:0xA|payload:0xABCD|raw:0xda120000]

        public Device(string name, Int32 address, Int32 rfNode, DeviceTypeEnm deviceType)
        {
            Name = name;
            Id = address;
            Node = rfNode;
            DeviceType = deviceType;
        }
        public String Name { get; set; }
        public Int32 Id { get; set; }
        public Command Cmd { get; set; }
        public MediaEnm Media { get; set; }
        public Int32 Payload { get; set; }
        public Int32 Value { get; set; }
        public DateTime ValueDate { get; set; }
        public DeviceStatus Status { get; set; }
        public DeviceTypeEnm DeviceType { get; set; }
        public String ValueStr { get; set; }
        public Int32 Node { get; set; }

        //logic related props
        public bool IncludeInUpdate { get; set; }
        public bool IncludeInEnviroment { get; set; }


        //Device inner logic
        public bool IsQueryable() => DeviceType == DeviceTypeEnm.Node ||
                                   DeviceType == DeviceTypeEnm.Light ||
                                   DeviceType == DeviceTypeEnm.LightArray ||
                                   DeviceType == DeviceTypeEnm.AcSwitch ||
                                   DeviceType == DeviceTypeEnm.AcTemp;

        public String GetCommand(Command command, Int32 value = 0)
        {
            string cmd = string.Empty;
            cmd += $"N{Node:X}";                    // node selection
            cmd += "D";                             // device prefix
            cmd += ((int)command).ToString("X");    // command
            cmd += Id.ToString("X2");               // device
            cmd += value.ToString("X4");            // value (only on set)
            return cmd;
        }

        public override string ToString() => $"[{Id:X2}:{Name}|{Value:X4}]";

        public enum DeviceStatus
        {
            Unknown = 0,
            Ok = 1,
            Awaiting = 2,
            Error = 3,
            NodeDown = 4
        }
        public enum DeviceTypeEnm
        {
            Unknown,
            Node,
            Light,
            LightArray,
            Temp,
            LightSensor,
            IrAppliance,
            AcSwitch,
            AcTemp,
            Analog,
            Humidity,
            Pressure,
            LedIndicator,
            Switch,
            IrReceiver
        }
        public enum MediaEnm
        {
            Serial = 0,
            Rf = 1,
            Ir = 2
        }
        public enum Command
        {
            QUERY = 0xA,
            VALUE = 0xB,
            SET = 0xC,
            EVENT = 0xD,
            STATUS = 0xE
        }


    }
}

File: HouzLink/HouzLink/wwwroot/src/Device.js
function Device(args) {
    this.id = args.id;
    this.name = args.name;
    this.cmd = args.cmd;
    this.payload = args.payload;
    this.value = args.value;
    this.valueDate = args.valueDate;
    this.status = undefined;
    this.deviceType = args.deviceType;
    this.valueStr = args.valueStr;
    this.node = args.node;
    this.includeInUpdate = args.includeInUpdate;
    this.includeInEnviroment = args.includeInEnviroment;

    //client fx
    this.elem = document.querySelector("*[device=\"" + this.id + "\"]");
    if (this.elem !== null)
        this
            .bindAction()
            .setStatus(args.status, true)
            .setValue(args.value, true);
    return this;
};

Device.prototype.bindAction = function () {
    this.elem.classList.add(enums.deviceType[this.deviceType]);
    switch (this.deviceType) {
    case 0: break; //Unknown
        case 2: //Light on/off
            this.elem.addEventListener("click", (function() {
                this.sendCommand(2, this.value === 0 ? 1 : 0);
            }).bind(this), false);
            break; 
        case 3: break;//LightArray

        case 10: //Humidity
        case 11: //Pressure
        case 4:  //Temp
            var sp = document.createElement("span");
            sp.innerHTML = this.value / 100;
            this.elem.appendChild(sp);

            this.elem.addEventListener("click", (function () {
                this.sendCommand(0);
            }).bind(this), false);
            break;   

        case 5: break;   //LightSensor
        case 6: break;   //IrAppliance
        case 7: break;   //AcSwitch
        case 8: break;   //AcTemp
        case 9: break;   //Analog
        case 12: break;  //LedIndicator
        case 13: break;  //Switch
        case 14: break;  //IrReceiver
    }
    return this;
};

Device.prototype.sendCommand = function (command, payload) {
    this.setStatus(2);
    switch (command) {
        case 0: //query
            ui.getText("device/" + this.id + "/query");
            break;
        case 2: //set
            ui.getText("device/" + this.id + "/set/" + payload);
    default:
    }

    return this;
};

Device.prototype.update = function (newVal) {
    return this
            .setStatus(newVal.status)
            .setValue(newVal.value);
};

Device.prototype.setStatus = function (newStatus, redraw) {
    if ((!redraw) && newStatus === this.status) return this;
    this.status = newStatus;
    const st = enums.deviceStatus[this.status];
    if (this.elem.classList.contains(st)) return this;
    for (let cls of enums.deviceStatus) this.elem.classList.remove(cls.toLowerCase());
    this.elem.classList.add(st);
    return this;
};

Device.prototype.setValue = function (newValue, redraw) {
    if ((!redraw) && newValue === this.value) return this;
    this.payload = this.value = newValue;
    switch (this.deviceType) {
        case 2: //light
            console.log("device.setValue [light]:", this.value);
            for (let cls of ["lightOn", "lightOff"]) this.elem.classList.remove(cls);
            this.elem.classList.add(this.value === 0 ? "lightOff" : "lightOn");
                break;
        case 10: //Humidity
        case 11: //Pressure
        case 4:  //Temp
            console.log("device.setValue [env]:", this.value);
            this.elem.querySelector("span").innerText = this.value;
    }
}
